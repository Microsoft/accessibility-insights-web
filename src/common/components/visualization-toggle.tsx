// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IToggle, IToggleProps, Toggle } from 'office-ui-fabric-react/lib/Toggle';
import * as React from 'react';

export interface IVisualizationToggleProps {
    checked: boolean;
    onClick: (event) => void;
    disabled?: boolean;
    label?: string;
    className?: string;
    visualizationName: string;
    componentRef?: React.RefObject<IToggle>;
    onBlur?: (ev) => void;
    onFocus?: (ev) => void;
}

export class VisualizationToggle extends React.Component<IVisualizationToggleProps> {
    public render(): JSX.Element {
        const props: IToggleProps = {
            checked: this.props.checked,
            onClick: this.props.onClick,
            disabled: this.props.disabled,
            label: this.props.label,
            className: this.props.className,
            onText: 'On',
            offText: 'Off',
            onAriaLabel: this.props.visualizationName,
            offAriaLabel: this.props.visualizationName,
            componentRef: this.props.componentRef,
            onBlur: this.props.onBlur,
            onFocus: this.props.onFocus,
        };

        return (
            <Toggle {...props}/>
        );
    }
}
