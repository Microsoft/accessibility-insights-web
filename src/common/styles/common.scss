// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import 'high-contrast-override.scss';
@import 'colors.scss';
@import 'fonts.scss';

$focus-ring-color: $neutral-100;

@mixin set-focus-ring-styles($property: 'border') {
    #{$property}: 1px solid $focus-ring-color;
    @include ms-high-contrast-override {
        #{$property}: 1px dashed windowtext !important;
    }
}

.ms-Fabric.is-focusVisible .ms-Nav-group .ms-nav-linkButton:focus::after {
    @include set-focus-ring-styles('border');
}

button::after {
    content: '';
    position: absolute;
}

.ms-Button.ms-Button--primary:not(.is-disabled) {
    background: $communication-primary;

    &:hover {
        background: $communication-shade-10;
    }
}

.ms-fontColor-neutralPrimary,
.ms-fontColor-neutralPrimary--hover:hover {
    color: $neutral-100 !important;
}

.ms-Fabric {
    color: $neutral-100;
}

.generic-toggle-component {
    margin-bottom: 4vh;
    .toggle-container {
        margin-bottom: 1vh;
        .toggle-name {
            display: inline-block;
            font-size: 17px;
            font-weight: bold;
        }

        .toggle {
            float: right;
        }
    }
}

.high-contrast-theme {
    .insights-link {
        &:hover {
            text-decoration: underline;
        }
    }
    .is-selected {
        .status-icon {
            color: $high-contrast-icon-color !important;
        }
    }
}

.insights-link {
    // marking important to override office fabric style
    color: $link !important;
    text-decoration: none;

    &:hover {
        color: $link-hover !important;
        text-decoration: underline;
    }
}

.ms-Spinner {
    .ms-Spinner-circle {
        border-top-color: $spinner-text;
    }

    .ms-Spinner-label {
        color: $spinner-text;
    }
}

$normalTitleFontWeight: 300 !default;

$pivotItemLeftBorderWidth: 3px !default;
$pivotItemBorderStyle: solid !default;

$detailsViewHeaderBarHeight: 40px !default;
$detailsViewNavPivotsHeight: 41px !default;
$detailsViewCommandBarHeight: 40px !default;
$detailsViewCommandBarBorderHeight: 1px !default;
$detailsViewLeftNavWidth: 200px !default;
$detailsViewTotalHeaderHeight: (#{$detailsViewCommandBarBorderHeight} + #{$detailsViewCommandBarHeight} + #{$detailsViewHeaderBarHeight});

$fastPassRightPanelMarginRight: 24px !default;
$fastPassRightPanelMarginLeft: 24px !default;
$fastPassRightPanelMarginTop: 24px !default;

@mixin ellipsedText {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}
