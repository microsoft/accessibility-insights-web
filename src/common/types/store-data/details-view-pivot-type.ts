// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export enum DetailsViewPivotType {
    fastPass = 0,
    assessment = 2,
}
