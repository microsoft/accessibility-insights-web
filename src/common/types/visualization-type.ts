// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export enum VisualizationType {
    Headings,
    Issues,
    Landmarks,
    TabStops,
    Color,
    HeadingsAssessment,
    ColorSensoryAssessment,
    LanguageAssessment,
    LandmarksAssessment,
    PageAssessment,
    RepetitiveContentAssessment,
    KeyboardInteraction,
    AudioVideOnly,
    ErrorsAssessment,
    TimedEventsAssessment,
    ParsingAssessment,
    PrerecordedMultimediaAssessment,
    LiveMultimediaAssessment,
    VisibleFocusOrderAssessment,
    ImagesAssessment,
    AdaptableContent,
    LinksAssessment,
    NativeWidgets,
    CustomWidgets,
    AutomatedChecks,
    SequenceAssessment,
    SemanticsAssessment,
    PointerMotionAssessment,
    ContrastAssessment,
    NeedsReview,
    AccessibleNames,
}
