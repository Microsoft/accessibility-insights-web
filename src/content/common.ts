// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as react from 'react';

import { link } from './link';
import { ContentCreator } from '../views/content/content-page';
import { GuidanceTitle as gt } from '../views/content/guidance-title';

export const create = ContentCreator(link);
export const React = react;
export const GuidanceTitle = gt;
