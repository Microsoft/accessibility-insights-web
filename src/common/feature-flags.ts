// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as _ from 'lodash';

import { FeatureFlagStoreData } from './types/store-data/feature-flag-store-data';

export class FeatureFlags {
    public static readonly logTelemetryToConsole = 'logTelemetryToConsole';
    public static readonly newAssessmentExperience = 'newAssessmentExperience';
    public static readonly showAllAssessments = 'showAllAssessments';
    public static readonly shadowDialog = 'shadowDialog';
    public static readonly exportResult = 'exportResult';
    public static readonly showAllFeatureFlags = 'showAllFeatureFlags';
    public static readonly scoping = 'scoping';
    public static readonly showBugFiling = 'showBugFiling';
    public static readonly showInstanceVisibility = 'showInstanceVisibility';
    public static readonly highContrastMode = 'highContrastMode';
}

export interface FeatureFlagDetail {
    id: string;
    defaultValue: boolean;
    displayableName: string;
    displayableDescription: string;
    isPreviewFeature: boolean;
    forceDefault: boolean;
}

export function getAllFeatureFlagDetails(): FeatureFlagDetail[] {
    return [
        {
            id: FeatureFlags.exportResult,
            defaultValue: true,
            displayableName: 'Export automated checks',
            displayableDescription: 'Allows you to export a summary of automated check results as a sharable HTML file.',
            isPreviewFeature: false,
            forceDefault: true,
        },
        {
            id: FeatureFlags.shadowDialog,
            defaultValue: false,
            displayableName: 'Improved dialog styling',
            displayableDescription:
                'Avoids styling problems in failure dialogs by rendering them in shadow DOM. ' +
                "(You'll need to refresh the target page to see the new dialog styling.)",
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.newAssessmentExperience,
            defaultValue: true,
            displayableName: 'New assessment experience',
            displayableDescription: 'Show new Assessment pivot and manual assisted test experiences under development.',
            isPreviewFeature: false,
            forceDefault: true,
        },
        {
            id: FeatureFlags.showAllAssessments,
            defaultValue: false,
            displayableName: 'Show all assessments',
            displayableDescription: 'Show all assessments, even the in-development ones',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.logTelemetryToConsole,
            defaultValue: false,
            displayableName: 'Log telemetry to console',
            displayableDescription: 'Write telemetry payload information to the developer tools console.',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.showAllFeatureFlags,
            defaultValue: false,
            displayableName: 'Show all feature flags',
            displayableDescription: 'Show all feature flags in the Preview Features panel.',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.scoping,
            defaultValue: false,
            displayableName: 'Scoping experience',
            displayableDescription: 'Enable scoping to limit scanning to selected portions of the webpage.',
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.showBugFiling,
            defaultValue: true,
            displayableName: 'Issue filing',
            displayableDescription: 'Enable File Issue buttons that allow you to create GitHub issues pre-populated with failure details.',
            isPreviewFeature: false,
            forceDefault: true,
        },
        {
            id: FeatureFlags.showInstanceVisibility,
            defaultValue: false,
            displayableName: 'Show instance visibility in assessment',
            displayableDescription:
                'Shows visibility of instances in assessment requirement lists. May impact performance. ' +
                "(You'll need to go to a different requirement and come back for it to take effect.)",
            isPreviewFeature: false,
            forceDefault: false,
        },
        {
            id: FeatureFlags.highContrastMode,
            defaultValue: false,
            displayableName: 'High contrast mode',
            displayableDescription: 'Show setting for high contrast mode under development',
            isPreviewFeature: true,
            forceDefault: false,
        },
    ];
}

export function getDefaultFeatureFlagValues(): FeatureFlagStoreData {
    const details: FeatureFlagDetail[] = getAllFeatureFlagDetails();
    const values: FeatureFlagStoreData = {};
    _.forEach(details, detail => {
        values[detail.id] = detail.defaultValue;
    });
    return values;
}

export function getForceDefaultFlags(): FeatureFlags[] {
    const details: FeatureFlagDetail[] = getAllFeatureFlagDetails();
    const forceDefaultFlags: FeatureFlags[] = [];
    _.forEach(details, detail => {
        if (detail.forceDefault) {
            forceDefaultFlags.push(detail.id);
        }
    });
    return forceDefaultFlags;
}
