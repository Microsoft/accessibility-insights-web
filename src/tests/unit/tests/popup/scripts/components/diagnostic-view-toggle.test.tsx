// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow, ShallowWrapper } from 'enzyme';
import * as Enzyme from 'enzyme';
import * as _ from 'lodash';
import { Link } from 'office-ui-fabric-react/lib/Link';
import { Spinner, SpinnerSize } from 'office-ui-fabric-react/lib/Spinner';
import { IToggle, Toggle, ToggleBase } from 'office-ui-fabric-react/lib/Toggle';
import * as React from 'react';
import * as TestUtils from 'react-dom/test-utils';
import { It, Mock, Times } from 'typemoq';

import { VisualizationToggle } from '../../../../../../common/components/visualization-toggle';
import {
    IVisualizationConfiguration,
    VisualizationConfigurationFactory,
} from '../../../../../../common/configs/visualization-configuration-factory';
import { FeatureFlags } from '../../../../../../common/feature-flags';
import { TelemetryEventSource } from '../../../../../../common/telemetry-events';
import { DetailsViewPivotType } from '../../../../../../common/types/details-view-pivot-type';
import { IVisualizationStoreData } from '../../../../../../common/types/store-data/ivisualization-store-data';
import { VisualizationType } from '../../../../../../common/types/visualization-type';
import { PopupActionMessageCreator } from '../../../../../../popup/scripts/actions/popup-action-message-creator';
import {
    DiagnosticViewToggle,
    DiagnosticViewToggleProps,
    DiagnosticViewToggleState,
} from '../../../../../../popup/scripts/components/diagnostic-view-toggle';
import { DiagnosticViewClickHandler } from '../../../../../../popup/scripts/handlers/diagnostic-view-toggle-click-handler';
import { ContentLink, ContentLinkDeps } from '../../../../../../views/content/content-link';
import { EventStubFactory } from '../../../../common/event-stub-factory';
import { ShortcutCommandsTestData } from '../../../../common/sample-test-data';
import { VisualizationStoreDataBuilder } from '../../../../common/visualization-store-data-builder';

describe('DiagnosticViewToggleTest', () => {
    const visualizationConfigurationFactory = new VisualizationConfigurationFactory();
    const testTelemetrySource: TelemetryEventSource = -1 as TelemetryEventSource;
    const eventStubFactory = new EventStubFactory();

    test('render spinner', () => {
        const type = VisualizationType.Headings;
        const data = new VisualizationStoreDataBuilder().with('scanning', 'headings').build();

        const props: DiagnosticViewToggleProps = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource)
            .setupVisualizationStoreData(data)
            .build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);

        assertSpinner(wrapper);
        assertDetailsViewLink(wrapper, type);
        assertShortcut(wrapper, type, props);
    });

    test('render toggle when scanning', () => {
        const type = VisualizationType.Headings;
        const data = new VisualizationStoreDataBuilder().with('scanning', 'landmarks').build();

        const props: DiagnosticViewToggleProps = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource)
            .setupVisualizationStoreData(data)
            .build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);

        assertVisualizationToggle(wrapper, type, data);
        assertDetailsViewLink(wrapper, type);
        assertShortcut(wrapper, type, props);
    });

    test('render toggle when not scanning', () => {
        const type = VisualizationType.Headings;

        const props: DiagnosticViewToggleProps = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);

        assertVisualizationToggle(wrapper, type, props.visualizationStoreData, false);
        assertDetailsViewLink(wrapper, type);
        assertShortcut(wrapper, type, props);
    });

    test('verify content link rendered when newAssessmentExperience is on', () => {
        const type = VisualizationType.Headings;

        const props: DiagnosticViewToggleProps = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource)
            .setupFeatureFlags({ [FeatureFlags.newAssessmentExperience]: true })
            .build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);

        assertVisualizationToggle(wrapper, type, props.visualizationStoreData, false);
        assertContentLink(wrapper, type);
        assertShortcut(wrapper, type, props);
    });

    test('verify details view link rendered when newAssessmentExperience is on, but the test does not have a guidance', () => {
        const type = VisualizationType.Issues;

        const props: DiagnosticViewToggleProps = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource)
            .setupFeatureFlags({ [FeatureFlags.newAssessmentExperience]: true })
            .build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);

        assertVisualizationToggle(wrapper, type, props.visualizationStoreData, false);
        assertDetailsViewLink(wrapper, type);
        assertShortcut(wrapper, type, props);
    });

    test('verify details view link will open fastpass when Assessment enabled', () => {
        const type = VisualizationType.Issues;
        const event = eventStubFactory.createKeypressEvent();
        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource);
        const props: DiagnosticViewToggleProps = propsBuilder
            .setupFeatureFlags({ [FeatureFlags.newAssessmentExperience]: true })
            .setupOpenDetailsViewCall(event, DetailsViewPivotType.fastPass)
            .build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);

        wrapper.find(Link).simulate('click', event);

        propsBuilder.verifyAll();
    });

    test('click the visualization toggle', () => {
        const type = VisualizationType.Headings;
        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupToggleVisualizationCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);

        wrapper.find(VisualizationToggle).simulate('click', event);

        propsBuilder.verifyAll();
    });

    test('click details view toggle when spinner is present', () => {
        const type = VisualizationType.Issues;
        const data = new VisualizationStoreDataBuilder().with('scanning', 'issues').build();

        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource)
            .setupVisualizationStoreData(data)
            .setupOpenDetailsViewCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);
        wrapper.find(Link).simulate('click', event);

        propsBuilder.verifyAll();
    });

    test('click details view toggle when toggle is present', () => {
        const type = VisualizationType.TabStops;
        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupOpenDetailsViewCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const wrapper = shallow(<DiagnosticViewToggle {...props} />);
        wrapper.find(Link).simulate('click', event);

        propsBuilder.verifyAll();
    });

    test('command not found', () => {
        const type = VisualizationType.Color;
        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupShortcutCommands([]);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const component = new DiagnosticViewToggle(props);

        const renderAction = () => component.render();

        const commandName = visualizationConfigurationFactory.getConfiguration(type).chromeCommand;
        expect(renderAction).toThrowError(`Cannot find command for name: ${commandName}`);
    });

    test('set focus when componentDidMount', () => {
        const type = VisualizationType.TabStops;
        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupOpenDetailsViewCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const wrapper = Enzyme.mount(<DiagnosticViewToggle {...props} />);
        wrapper.setState({
            isFocused: true,
        });
        const toggle = wrapper.find(VisualizationToggle).props().componentRef.current;
        jest.spyOn(toggle, 'focus');

        wrapper.instance().componentDidMount();

        expect(toggle.focus).toHaveBeenCalledTimes(1);
    });

    test('componentDidUpdate', () => {
        const type = VisualizationType.TabStops;
        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupOpenDetailsViewCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const wrapper = Enzyme.mount(<DiagnosticViewToggle {...props} />);
        wrapper.setState({
            isFocused: true,
        });
        const toggle = wrapper.find(VisualizationToggle).props().componentRef.current;
        jest.spyOn(toggle, 'focus');

        wrapper.instance().componentDidUpdate(props, props);

        expect(toggle.focus).toHaveBeenCalledTimes(1);
    });

    test('onFocusHandler', () => {
        const type = VisualizationType.TabStops;
        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupOpenDetailsViewCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const component = React.createElement(DiagnosticViewToggle, props);

        const testObject = TestUtils.renderIntoDocument(component);

        (testObject as any)._isMounted = true;

        const onFocusHandlerFunction = (testObject as any).onFocusHandler;
        onFocusHandlerFunction();

        expect((testObject as any).state.isFocused).toBeTruthy();
    });

    test('onBlurHandler', () => {
        const type = VisualizationType.TabStops;
        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupOpenDetailsViewCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const component = React.createElement(DiagnosticViewToggle, props);

        const testObject = TestUtils.renderIntoDocument(component);

        (testObject as any)._isMounted = true;

        const onBlurHandlerFunction = (testObject as any).onBlurHandler;
        onBlurHandlerFunction();

        expect((testObject as any).state.isFocused).toBeFalsy();
    });

    test('addUserEventListener', () => {
        const type = VisualizationType.TabStops;
        const event = eventStubFactory.createKeypressEvent();

        const propsBuilder = new DiagnosticViewTogglePropsBuilder(type, testTelemetrySource).setupOpenDetailsViewCall(event);

        const props: DiagnosticViewToggleProps = propsBuilder.build();

        const component = React.createElement(DiagnosticViewToggle, props);

        const testObject = TestUtils.renderIntoDocument(component);

        (testObject as any)._userEventListenerAdded = false;

        const addUserEventListenerFunction = (testObject as any).addUserEventListener;
        addUserEventListenerFunction();

        expect((testObject as any)._userEventListenerAdded).toBeTruthy();
        propsBuilder.addUserListenerVerifyAll();
    });

    function assertSpinner(wrapper: ShallowWrapper<DiagnosticViewToggleProps, DiagnosticViewToggleState>): void {
        const spinner = wrapper.find(Spinner);

        expect(spinner).toBeDefined();
        expect(spinner.props().size).toEqual(SpinnerSize.small);
    }

    function assertDetailsViewLink(
        wrapper: ShallowWrapper<DiagnosticViewToggleProps, DiagnosticViewToggleState>,
        type: VisualizationType,
    ): void {
        const detailsViewLink = wrapper.find(Link);

        const configuration = visualizationConfigurationFactory.getConfiguration(type);
        const displayableData = configuration.displayableData;

        expect(detailsViewLink).toBeDefined();
        expect(detailsViewLink.children().text()).toEqual(displayableData.linkToDetailsViewText);
        expect(detailsViewLink.props().href).toEqual('#');
    }

    function assertContentLink(
        wrapper: ShallowWrapper<DiagnosticViewToggleProps, DiagnosticViewToggleState>,
        type: VisualizationType,
    ): void {
        const detailsViewLink = wrapper.find(ContentLink);

        const configuration = visualizationConfigurationFactory.getConfiguration(type);
        const displayableData = configuration.displayableData;

        expect(detailsViewLink.props().linkText).toEqual(displayableData.linkToDetailsViewText);
    }

    function assertShortcut(
        wrapper: ShallowWrapper<DiagnosticViewToggleProps, DiagnosticViewToggleState>,
        type: VisualizationType,
        props: DiagnosticViewToggleProps,
    ): void {
        const shortcutDiv = wrapper.findWhere(node => node.hasClass('shortcut-label'));

        expect(shortcutDiv).toBeDefined();

        const commandName = visualizationConfigurationFactory.getConfiguration(type).chromeCommand;
        const expectedCommand = _.find(props.shortcutCommands, command => command.name === commandName);

        expect(shortcutDiv.text()).toEqual(expectedCommand.shortcut);
    }

    function assertVisualizationToggle(
        wrapper: ShallowWrapper<DiagnosticViewToggleProps, DiagnosticViewToggleState>,
        type: VisualizationType,
        data: IVisualizationStoreData,
        isDisabled: boolean = true,
    ): void {
        const visualizationToggle = wrapper.find(VisualizationToggle);

        expect(visualizationToggle).toBeDefined();

        const configuration = visualizationConfigurationFactory.getConfiguration(type);
        const scanData = configuration.getStoreData(data.tests);

        expect(visualizationToggle.props().checked).toEqual(scanData.enabled);
        expect(visualizationToggle.props().disabled).toEqual(isDisabled);
    }
});

class DiagnosticViewTogglePropsBuilder {
    private type: VisualizationType;
    private data: IVisualizationStoreData = new VisualizationStoreDataBuilder().build();
    private visualizationConfigurationFactory = new VisualizationConfigurationFactory();
    private defaultVisualizationConfigurationFactoryMock = Mock.ofType(VisualizationConfigurationFactory);
    private actionMessageCreatorMock = Mock.ofType(PopupActionMessageCreator);
    private clickHandlerMock = Mock.ofType(DiagnosticViewClickHandler);
    private telemetrySource: TelemetryEventSource;
    private shortcutCommands: chrome.commands.Command[] = ShortcutCommandsTestData;
    private querySelectorMock = Mock.ofInstance(selector => {});
    private addEventListenerMock = Mock.ofInstance((e, ev) => {});
    private featureFlags: IDictionaryStringTo<boolean> = {};
    private deps: ContentLinkDeps = {} as ContentLinkDeps;
    private configurationStub: IVisualizationConfiguration;

    constructor(type: VisualizationType, telemetrySource: TelemetryEventSource) {
        this.type = type;
        this.telemetrySource = telemetrySource;
    }

    public setupShortcutCommands(shortcutCommands: chrome.commands.Command[]): DiagnosticViewTogglePropsBuilder {
        this.shortcutCommands = shortcutCommands;
        return this;
    }

    public setupVisualizationStoreData(data: IVisualizationStoreData): DiagnosticViewTogglePropsBuilder {
        this.data = data;
        return this;
    }

    public setupOpenDetailsViewCall(event, pivot = DetailsViewPivotType.allTest): DiagnosticViewTogglePropsBuilder {
        this.actionMessageCreatorMock
            .setup(ac => ac.openDetailsView(event, this.type, this.telemetrySource, pivot))
            .verifiable(Times.once());

        return this;
    }

    public setupToggleVisualizationCall(event): DiagnosticViewTogglePropsBuilder {
        this.clickHandlerMock.setup(ch => ch.toggleVisualization(this.data, this.type, event)).verifiable(Times.once());

        return this;
    }

    public setupFeatureFlags(featureFlags: IDictionaryStringTo<boolean>): DiagnosticViewTogglePropsBuilder {
        this.featureFlags = featureFlags;
        return this;
    }

    public build(): DiagnosticViewToggleProps {
        this.querySelectorMock
            .setup(q => q('.feedback-collapse-menu-button'))
            .returns(s => null)
            .verifiable(Times.once());

        this.addEventListenerMock
            .setup(a => a('keydown', It.isAny()))
            .returns((e, ev) => true)
            .verifiable(Times.once());

        const domMock = {
            querySelector: this.querySelectorMock.object,
            addEventListener: this.addEventListenerMock.object,
        };

        this.defaultVisualizationConfigurationFactoryMock
            .setup(v => v.getConfiguration(It.isAny()))
            .returns(type => this.configurationStub || this.visualizationConfigurationFactory.getConfiguration(type))
            .verifiable();

        const props: DiagnosticViewToggleProps = {
            deps: this.deps,
            featureFlags: this.featureFlags,
            dom: domMock as any,
            type: this.type,
            visualizationConfigurationFactory: this.visualizationConfigurationFactory,
            visualizationStoreData: this.data,
            actionMessageCreator: this.actionMessageCreatorMock.object,
            clickHandler: this.clickHandlerMock.object,
            shortcutCommands: this.shortcutCommands,
            telemetrySource: this.telemetrySource,
        };

        return props;
    }

    public verifyAll(): void {
        this.actionMessageCreatorMock.verifyAll();
        this.clickHandlerMock.verifyAll();
    }

    public addUserListenerVerifyAll(): void {
        this.querySelectorMock.verifyAll();
        this.addEventListenerMock.verifyAll();
    }
}
