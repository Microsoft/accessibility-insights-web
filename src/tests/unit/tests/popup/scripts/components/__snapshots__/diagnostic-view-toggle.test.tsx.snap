// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiagnosticViewToggleTest renders details view link when the test does not have a guidance 1`] = `
<div>
  <div
    className="ms-Grid-row view-toggle-row"
  >
    <div
      className="ms-Grid-col ms-sm7"
    >
      <div
        className="ms-fontColor-neutralPrimary ms-fontWeight-semibold ms-fontSize-mPlus"
      >
        Automated checks
      </div>
    </div>
    <div
      className="ms-Grid-col ms-sm5 view-toggle-col"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <VisualizationToggle
        checked={false}
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        visualizationName="Automated checks"
      />
    </div>
  </div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-sm8"
    >
      <StyledLinkBase
        className="insights-link"
        href="#"
        onClick={[Function]}
      >
        List view and filtering
      </StyledLinkBase>
    </div>
    <div
      className="ms-Grid-col ms-sm4 shortcut-label"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <div
        className="ms-font-xs"
      >
        Ctrl+Shift+1
      </div>
    </div>
  </div>
</div>
`;

exports[`DiagnosticViewToggleTest renders spinner while scanning 1`] = `
<div>
  <div
    className="ms-Grid-row view-toggle-row"
  >
    <div
      className="ms-Grid-col ms-sm7"
    >
      <div
        className="ms-fontColor-neutralPrimary ms-fontWeight-semibold ms-fontSize-mPlus"
      >
        Headings
      </div>
    </div>
    <div
      className="ms-Grid-col ms-sm5 view-toggle-col"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <StyledSpinnerBase
        componentRef={[Function]}
        size={1}
      />
    </div>
  </div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-sm8"
    >
      <ContentLink
        deps={Object {}}
        linkText="How to test headings"
        reference={[Function]}
      />
    </div>
    <div
      className="ms-Grid-col ms-sm4 shortcut-label"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <div
        className="ms-font-xs"
      >
        Ctrl+Shift+3
      </div>
    </div>
  </div>
</div>
`;

exports[`DiagnosticViewToggleTest renders toggle when not scanning 1`] = `
<div>
  <div
    className="ms-Grid-row view-toggle-row"
  >
    <div
      className="ms-Grid-col ms-sm7"
    >
      <div
        className="ms-fontColor-neutralPrimary ms-fontWeight-semibold ms-fontSize-mPlus"
      >
        Headings
      </div>
    </div>
    <div
      className="ms-Grid-col ms-sm5 view-toggle-col"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <VisualizationToggle
        checked={false}
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        visualizationName="Headings"
      />
    </div>
  </div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-sm8"
    >
      <ContentLink
        deps={Object {}}
        linkText="How to test headings"
        reference={[Function]}
      />
    </div>
    <div
      className="ms-Grid-col ms-sm4 shortcut-label"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <div
        className="ms-font-xs"
      >
        Ctrl+Shift+3
      </div>
    </div>
  </div>
</div>
`;

exports[`DiagnosticViewToggleTest renders toggle when scanning for a different visualization 1`] = `
<div>
  <div
    className="ms-Grid-row view-toggle-row"
  >
    <div
      className="ms-Grid-col ms-sm7"
    >
      <div
        className="ms-fontColor-neutralPrimary ms-fontWeight-semibold ms-fontSize-mPlus"
      >
        Headings
      </div>
    </div>
    <div
      className="ms-Grid-col ms-sm5 view-toggle-col"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <VisualizationToggle
        checked={false}
        componentRef={
          Object {
            "current": null,
          }
        }
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        visualizationName="Headings"
      />
    </div>
  </div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-sm8"
    >
      <ContentLink
        deps={Object {}}
        linkText="How to test headings"
        reference={[Function]}
      />
    </div>
    <div
      className="ms-Grid-col ms-sm4 shortcut-label"
      style={
        Object {
          "float": "right",
        }
      }
    >
      <div
        className="ms-font-xs"
      >
        Ctrl+Shift+3
      </div>
    </div>
  </div>
</div>
`;
